File: trace_analyzer.pl
#!/usr/bin/perl

# Trace Log Analyzer - High Performance Perl Version
# Optimized for very large trace files (100MB+)
#
# Usage: perl trace_analyzer.pl <log_file> <entry_pattern> <exit_pattern> <threshold_seconds>
# Example: perl trace_analyzer.pl trace.log "doRequest ENTRY" "doRequest RETURN" 3

use strict;
use warnings;
use Time::HiRes qw(time);

# Configuration
my $PROGRESS_INTERVAL = 5000;  # Show progress every N lines
my $MATCH_PROGRESS_INTERVAL = 500;  # Show matching progress every N matches

sub usage {
    print "Usage: $0 <log_file> <entry_pattern> <exit_pattern> <threshold_seconds>\n\n";
    print "Example:\n";
    print "  $0 trace.log \"doRequest ENTRY\" \"doRequest RETURN\" 3\n\n";
    print "Arguments:\n";
    print "  log_file        - Path to the trace log file\n";
    print "  entry_pattern   - Pattern to search for entry points\n";
    print "  exit_pattern    - Pattern to search for exit points\n";
    print "  threshold_seconds - Time threshold in seconds to report slow operations\n";
    exit 1;
}

sub format_size {
    my $bytes = shift;
    if ($bytes >= 1024*1024*1024) {
        return sprintf("%.1f GB", $bytes / (1024*1024*1024));
    } elsif ($bytes >= 1024*1024) {
        return sprintf("%.1f MB", $bytes / (1024*1024));
    } elsif ($bytes >= 1024) {
        return sprintf("%.1f KB", $bytes / 1024);
    } else {
        return "$bytes bytes";
    }
}

sub show_progress {
    my ($current, $total, $task, $start_time) = @_;
    my $percent = $total > 0 ? ($current * 100.0 / $total) : 0;
    my $elapsed = time() - $start_time;
    my $rate = $elapsed > 0 ? $current / $elapsed : 0;
    my $eta = $rate > 0 ? ($total - $current) / $rate : 0;
    
    printf "\r%s: %d/%d (%.1f%%) - %.1f lines/sec - ETA: %.0fs", 
           $task, $current, $total, $percent, $rate, $eta;
}

sub clear_progress {
    print "\r" . " " x 80 . "\r";
}

sub parse_timestamp {
    my $timestamp_str = shift;
    
    # Parse format: [9/12/25, 13:25:29:271 CDT]
    if ($timestamp_str =~ /\[(\d+)\/(\d+)\/(\d+),\s*(\d+):(\d+):(\d+):(\d+)\s+[A-Z]{3}\]/) {
        my ($month, $day, $year, $hour, $min, $sec, $ms) = ($1, $2, $3, $4, $5, $6, $7);
        $year = 2000 + $year if $year < 100;
        
        # Convert to seconds since epoch (simplified calculation)
        # This is approximate but consistent for duration calculations
        my $days_since_epoch = ($year - 1970) * 365 + int(($year - 1969) / 4);  # Rough estimate
        $days_since_epoch += ($month - 1) * 30.5 + $day;  # Rough month calculation
        
        my $seconds = $days_since_epoch * 86400 + $hour * 3600 + $min * 60 + $sec;
        return $seconds + ($ms / 1000.0);
    }
    return 0;
}

sub extract_method_name {
    my $pattern = shift;
    my ($method) = split /\s+/, $pattern, 2;
    return $method || "unknown";
}

# Check arguments
if (@ARGV != 4) {
    usage();
}

my ($log_file, $entry_pattern, $exit_pattern, $threshold) = @ARGV;

# Validate arguments
unless (-f $log_file) {
    die "Error: Log file '$log_file' not found\n";
}

unless ($threshold =~ /^[\d.]+$/) {
    die "Error: Threshold must be a number\n";
}

# Get file info
my $file_size = -s $log_file;
print "Analyzing trace file: $log_file\n";
print "File size: " . format_size($file_size) . "\n";
print "Entry pattern: '$entry_pattern'\n";
print "Exit pattern: '$exit_pattern'\n";
print "Threshold: ${threshold} seconds\n";
print "=" x 80 . "\n";

my $method_name = extract_method_name($entry_pattern);

# Data structures for fast lookups
my %thread_entries = ();  # Hash of hashes: thread_id -> method -> [timestamp, raw_timestamp, line_data]
my @results = ();         # Array of results exceeding threshold
my $total_pairs = 0;
my $slow_operations = 0;
my $entry_count = 0;
my $exit_count = 0;

# Phase 1: Process the log file
print "Phase 1: Reading and processing log file...\n";
my $start_time = time();
my $line_count = 0;

open my $fh, '<', $log_file or die "Cannot open $log_file: $!\n";

while (my $line = <$fh>) {
    $line_count++;
    
    # Show progress
    if ($line_count % $PROGRESS_INTERVAL == 0) {
        show_progress($line_count, 0, "Processing", $start_time);  # Don't know total lines yet
    }
    
    chomp $line;
    next unless $line;
    
    # Remove line numbers if present
    $line =~ s/^\d+\|//;
    
    # Extract timestamp and thread ID using regex (much faster than multiple operations)
    next unless $line =~ /(\[[^\]]+\])\s+([a-fA-F0-9]+)\s+/;
    my ($timestamp_str, $thread_id) = ($1, $2);
    
    if (index($line, $entry_pattern) >= 0) {
        # Process entry
        my $timestamp = parse_timestamp($timestamp_str);
        $thread_entries{$thread_id} = {} unless exists $thread_entries{$thread_id};
        $thread_entries{$thread_id}{$method_name} = [$timestamp, $timestamp_str, $line];
        $entry_count++;
        
    } elsif (index($line, $exit_pattern) >= 0) {
        # Process exit - try to match with existing entry
        if (exists $thread_entries{$thread_id} && 
            exists $thread_entries{$thread_id}{$method_name}) {
            
            my ($entry_time, $entry_timestamp_str, $entry_line) = 
                @{$thread_entries{$thread_id}{$method_name}};
            
            my $exit_time = parse_timestamp($timestamp_str);
            my $duration = $exit_time - $entry_time;
            
            $total_pairs++;
            
            if ($duration >= $threshold) {
                $slow_operations++;
                push @results, {
                    thread_id => $thread_id,
                    method => $method_name,
                    duration => $duration,
                    entry_time => $entry_timestamp_str,
                    exit_time => $timestamp_str,
                    entry_line => $entry_line,
                    exit_line => $line
                };
            }
            
            # Remove the matched entry to prevent duplicate matching
            delete $thread_entries{$thread_id}{$method_name};
            delete $thread_entries{$thread_id} unless keys %{$thread_entries{$thread_id}};
        }
        $exit_count++;
    }
}

close $fh;
clear_progress();

my $processing_time = time() - $start_time;

print "Phase 1 complete:\n";
print "  Total lines processed: $line_count\n";
print "  Found $entry_count entries and $exit_count exits\n";
print "  Matched $total_pairs entry/exit pairs\n";
print "  Processing time: " . sprintf("%.1f", $processing_time) . " seconds\n";
print "  Processing rate: " . sprintf("%.0f", $line_count / $processing_time) . " lines/second\n";

# Check if we found any patterns
if ($entry_count == 0 && $exit_count == 0) {
    print "\nNo matching patterns found in the log file!\n";
    print "Please verify your patterns match the log format exactly.\n";
    exit 1;
}

# Phase 2: Sort and display results
print "\nPhase 2: Sorting results...\n";

# Sort results by duration (descending)
@results = sort { $b->{duration} <=> $a->{duration} } @results;

print "=" x 80 . "\n";
print "\nAnalysis Results:\n";
print "Total matched entry/exit pairs: $total_pairs\n";

if ($slow_operations > 0) {
    print "\nOperations exceeding ${threshold} seconds threshold:\n";
    print "=" x 120 . "\n";
    printf "%-10s %-30s %-12s %-25s %-25s\n", 
           "Thread ID", "Method", "Duration (s)", "Entry Time", "Exit Time";
    print "=" x 120 . "\n";
    
    foreach my $result (@results) {
        # Clean timestamps for display
        my $clean_entry = $result->{entry_time};
        my $clean_exit = $result->{exit_time};
        $clean_entry =~ s/^\[|\]$//g;
        $clean_exit =~ s/^\[|\]$//g;
        
        printf "%-10s %-30s %-12.3f %-25s %-25s\n",
               $result->{thread_id}, $result->{method}, $result->{duration},
               $clean_entry, $clean_exit;
    }
    
    # Calculate statistics
    my @durations = map { $_->{duration} } @results;
    my $min_duration = $durations[-1];  # Last element (smallest after sorting desc)
    my $max_duration = $durations[0];   # First element (largest)
    my $sum = 0;
    $sum += $_ for @durations;
    my $avg_duration = @durations > 0 ? $sum / @durations : 0;
    
    print "\nTiming Statistics (for operations above threshold):\n";
    printf "  Minimum duration: %.3f seconds\n", $min_duration;
    printf "  Maximum duration: %.3f seconds\n", $max_duration;
    printf "  Average duration: %.3f seconds\n", $avg_duration;
} else {
    print "\nNo operations found exceeding ${threshold} seconds threshold.\n";
}

# Show unmatched entries
my $unmatched_count = 0;
foreach my $thread (keys %thread_entries) {
    $unmatched_count += keys %{$thread_entries{$thread}};
}

if ($unmatched_count > 0) {
    print "\nWarning: $unmatched_count unmatched entry points found (no corresponding exits)\n";
}

my $total_time = time() - $start_time;
print "\nAnalysis complete.\n";
print "Total processing time: " . sprintf("%.1f", $total_time) . " seconds\n";
print "Overall rate: " . sprintf("%.0f", $line_count / $total_time) . " lines/second\n";

if ($file_size > 50*1024*1024) {  # For files > 50MB
    my $mb_per_sec = ($file_size / (1024*1024)) / $total_time;
    printf "File processing rate: %.1f MB/second\n", $mb_per_sec;
}

print "\nPerformance Summary:\n";
printf "  File size: %s\n", format_size($file_size);
printf "  Processing time: %.1f seconds\n", $total_time;
printf "  Memory usage: Minimal (streaming processing)\n";
printf "  Algorithm: Single-pass O(n) with hash lookups\n";

File: trace_analyzer.py
#!/usr/bin/env python3
"""
Trace Log Analyzer

This script parses trace log files to track thread entry and exit points,
calculate timing differences, and report methods that exceed a specified threshold.

Usage:
    python3 trace_analyzer.py <log_file> --entry "ENTRY_STRING" --exit "EXIT_STRING" --threshold SECONDS

Example:
    python3 trace_analyzer.py trace.log --entry "doRequest ENTRY" --exit "doRequest RETURN" --threshold 3
"""

import re
import argparse
from datetime import datetime
from typing import Dict, List, Tuple, Optional
import sys


class TraceEntry:
    def __init__(self, timestamp_str: str, thread_id: str, method_name: str, entry_type: str, raw_line: str):
        self.timestamp_str = timestamp_str
        self.thread_id = thread_id
        self.method_name = method_name
        self.entry_type = entry_type
        self.raw_line = raw_line
        self.timestamp = self._parse_timestamp(timestamp_str)
    
    def _parse_timestamp(self, timestamp_str: str) -> datetime:
        """Parse timestamp from format: [9/12/25, 13:25:29:271 CDT]"""
        # Remove brackets and timezone
        clean_ts = timestamp_str.strip('[]').rsplit(' ', 1)[0]
        # Parse the datetime
        return datetime.strptime(clean_ts, "%m/%d/%y, %H:%M:%S:%f")


class TraceAnalyzer:
    def __init__(self, entry_pattern: str, exit_pattern: str, threshold_seconds: float):
        self.entry_pattern = entry_pattern
        self.exit_pattern = exit_pattern
        self.threshold_seconds = threshold_seconds
        self.trace_pattern = re.compile(
            r'\[([^\]]+)\]\s+(\w+)\s+id=\w+\s+[\w.]+\s+[><]\s+(.+)'
        )
        # Track open entries by thread_id and method_name
        self.open_entries: Dict[str, Dict[str, TraceEntry]] = {}
        self.completed_pairs: List[Tuple[TraceEntry, TraceEntry, float]] = []
        
    def parse_line(self, line: str) -> Optional[TraceEntry]:
        """Parse a single log line and extract trace information"""
        line = line.strip()
        if not line:
            return None
            
        # Remove line numbers if present (like "1|", "2|", etc.)
        if '|' in line and line.split('|')[0].isdigit():
            line = '|'.join(line.split('|')[1:])
            
        match = self.trace_pattern.match(line)
        if not match:
            return None
            
        timestamp_str, thread_id, rest_of_line = match.groups()
        
        # Extract method name - it's typically the last word before ENTRY/RETURN
        parts = rest_of_line.split()
        if len(parts) < 2:
            return None
            
        # Look for our entry/exit patterns
        if self.entry_pattern in rest_of_line:
            method_name = self._extract_method_name(rest_of_line, self.entry_pattern)
            return TraceEntry(timestamp_str, thread_id, method_name, 'ENTRY', line)
        elif self.exit_pattern in rest_of_line:
            method_name = self._extract_method_name(rest_of_line, self.exit_pattern)
            return TraceEntry(timestamp_str, thread_id, method_name, 'EXIT', line)
            
        return None
        
    def _extract_method_name(self, line: str, pattern: str) -> str:
        """Extract method name from the pattern"""
        # Extract the method name from the pattern itself
        # For "doRequest ENTRY" -> extract "doRequest"
        # For "doRequest RETURN" -> extract "doRequest"
        words = pattern.split()
        if words:
            return words[0]
        return "unknown"
        
    def process_entry(self, entry: TraceEntry):
        """Process an entry trace"""
        if entry.thread_id not in self.open_entries:
            self.open_entries[entry.thread_id] = {}
            
        # Store the entry, keyed by method name
        self.open_entries[entry.thread_id][entry.method_name] = entry
        
    def process_exit(self, exit_entry: TraceEntry):
        """Process an exit trace and match with corresponding entry"""
        thread_id = exit_entry.thread_id
        method_name = exit_entry.method_name
        
        if (thread_id in self.open_entries and 
            method_name in self.open_entries[thread_id]):
            
            entry_trace = self.open_entries[thread_id][method_name]
            
            # Calculate time difference in seconds
            time_diff = (exit_entry.timestamp - entry_trace.timestamp).total_seconds()
            
            self.completed_pairs.append((entry_trace, exit_entry, time_diff))
            
            # Remove the matched entry
            del self.open_entries[thread_id][method_name]
            if not self.open_entries[thread_id]:
                del self.open_entries[thread_id]
                
    def analyze_file(self, filename: str):
        """Analyze a trace log file"""
        print(f"Analyzing trace file: {filename}")
        print(f"Looking for entry pattern: '{self.entry_pattern}'")
        print(f"Looking for exit pattern: '{self.exit_pattern}'")
        print(f"Threshold: {self.threshold_seconds} seconds")
        print("-" * 80)
        
        try:
            with open(filename, 'r') as file:
                line_num = 0
                for line in file:
                    line_num += 1
                    entry = self.parse_line(line)
                    
                    if entry:
                        if entry.entry_type == 'ENTRY':
                            self.process_entry(entry)
                        elif entry.entry_type == 'EXIT':
                            self.process_exit(entry)
                            
        except FileNotFoundError:
            print(f"Error: File '{filename}' not found")
            sys.exit(1)
        except Exception as e:
            print(f"Error reading file '{filename}': {e}")
            sys.exit(1)
            
    def report_results(self):
        """Generate and display the analysis results"""
        print(f"\nAnalysis Results:")
        print(f"Total matched entry/exit pairs: {len(self.completed_pairs)}")
        
        # Filter pairs that exceed the threshold
        slow_operations = [
            (entry, exit, duration) for entry, exit, duration in self.completed_pairs
            if duration >= self.threshold_seconds
        ]
        
        if slow_operations:
            print(f"\nOperations exceeding {self.threshold_seconds} seconds threshold:")
            print("-" * 120)
            print(f"{'Thread ID':<10} {'Method':<30} {'Duration (s)':<12} {'Entry Time':<25} {'Exit Time':<25}")
            print("-" * 120)
            
            for entry, exit, duration in sorted(slow_operations, key=lambda x: x[2], reverse=True):
                print(f"{entry.thread_id:<10} {entry.method_name:<30} {duration:<12.3f} "
                      f"{entry.timestamp.strftime('%H:%M:%S.%f')[:-3]:<25} "
                      f"{exit.timestamp.strftime('%H:%M:%S.%f')[:-3]:<25}")
        else:
            print(f"\nNo operations found exceeding {self.threshold_seconds} seconds threshold.")
            
        # Report unmatched entries
        unmatched_count = sum(len(methods) for methods in self.open_entries.values())
        if unmatched_count > 0:
            print(f"\nWarning: {unmatched_count} unmatched entry points found (no corresponding exits)")
            
        # Show summary statistics
        if self.completed_pairs:
            durations = [duration for _, _, duration in self.completed_pairs]
            print(f"\nTiming Statistics:")
            print(f"  Minimum duration: {min(durations):.3f} seconds")
            print(f"  Maximum duration: {max(durations):.3f} seconds")
            print(f"  Average duration: {sum(durations)/len(durations):.3f} seconds")


def main():
    parser = argparse.ArgumentParser(
        description="Analyze trace logs for entry/exit timing patterns",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Analyze doRequest timing with 3 second threshold
  python3 trace_analyzer.py trace.log --entry "doRequest ENTRY" --exit "doRequest RETURN" --threshold 3
  
  # Analyze any method with custom patterns
  python3 trace_analyzer.py trace.log --entry "ENTRY" --exit "RETURN" --threshold 1.5
        """
    )
    
    parser.add_argument('logfile', help='Path to the trace log file')
    parser.add_argument('--entry', required=True, 
                       help='Entry pattern to search for (e.g., "doRequest ENTRY")')
    parser.add_argument('--exit', required=True,
                       help='Exit pattern to search for (e.g., "doRequest RETURN")')
    parser.add_argument('--threshold', type=float, required=True,
                       help='Time threshold in seconds to report slow operations')
    
    args = parser.parse_args()
    
    analyzer = TraceAnalyzer(args.entry, args.exit, args.threshold)
    analyzer.analyze_file(args.logfile)
    analyzer.report_results()


if __name__ == "__main__":
    main()
